credentials/tls/certprovider/pemfile/watcher.go

```diff
@@ -0,0 +1,249 @@
+	CertRefreshDuration time.Duration
+	// RootRefreshDuration is the amount of time the plugin waits before
+	// checking for updates in the specified root file.
+	// Optional. If not set, a default value (2 hour) will be used.
+	RootRefreshDuration time.Duration
+}
+
+// NewProvider returns a new certificate provider plugin that is configured to
+// watch the PEM files specified in the passed in options.
+func NewProvider(o Options) (certprovider.Provider, error) {
+	if o.CertFile == "" && o.KeyFile == "" && o.RootFile == "" {
+		return nil, fmt.Errorf("pemfile: at least one credential file needs to be specified")
+	}
+	if keySpecified, certSpecified := o.KeyFile != "", o.CertFile != ""; keySpecified != certSpecified {
+		return nil, fmt.Errorf("pemfile: private key file and identity cert file should be both specified or not specified")
+	}
+	if o.CertRefreshDuration == 0 {
+		o.CertRefreshDuration = defaultCertRefreshDuration
+	}
+	if o.RootRefreshDuration == 0 {
+		o.RootRefreshDuration = defaultRootRefreshDuration
+	}
+
+	provider := &watcher{opts: o}
+	if o.CertFile != "" && o.KeyFile != "" {
+		provider.identityDistributor = newDistributor()
+	}
+	if o.RootFile != "" {
+		provider.rootDistributor = newDistributor()
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	provider.cancel = cancel
+	go provider.run(ctx)
+
+	return provider, nil
+}
+
+// watcher is a certificate provider plugin that implements the
+// certprovider.Provider interface. It watches a set of certificate and key
+// files and provides the most up-to-date key material for consumption by
+// credentials implementation.
+type watcher struct {
+	identityDistributor distributor
+	rootDistributor     distributor
+	opts                Options
+	certFileContents    []byte
+	keyFileContents     []byte
+	rootFileContents    []byte
+	cancel              context.CancelFunc
+}
+
+// distributor wraps the methods on certprovider.Distributor which are used by
+// the plugin. This is very useful in tests which need to know exactly when the
+// plugin updates its key material.
+type distributor interface {
+	KeyMaterial(ctx context.Context) (*certprovider.KeyMaterial, error)
+	Set(km *certprovider.KeyMaterial, err error)
+	Stop()
+}
+
+// updateIdentityDistributor checks if the cert/key files that the plugin is
+// watching have changed, and if so, reads the new contents and updates the
+// identityDistributor with the new key material.
+//
+// Skips updates when file reading or parsing fails.
+// TODO(easwars): Retry with limit (on the number of retries or the amount of
+// time) upon failures.
+func (w *watcher) updateIdentityDistributor() {
+	if w.identityDistributor == nil {
+		return
+	}
+
+	certFileContents, err := ioutil.ReadFile(w.opts.CertFile)
+	if err != nil {
+		logger.Warningf("certFile (%s) read failed: %v", w.opts.CertFile, err)
+		return
+	}
+	keyFileContents, err := ioutil.ReadFile(w.opts.KeyFile)
```

Comment:
I am a bit confused why we are doing a duplicate content check here. Is the main purpose to avoid updating the distributor if the contents are same?
I am not sure if this performance check is worth to do, because before, we will just do one read from disk for each files, and now we will need to do twice. Disk I/Os are usually more expensive than memory operations, unless we are doing some heavy operations in distributor...Even worse, in some extreme scenarios, the certificates could be large(since it's actually chain of certs), and it might take some more time to read them and compare them per bit.
It might be better if we can check the cert we got from `tls.LoadX509KeyPair` directly. Is it possible?
Also, I was under the impression that the distributor will check the duplicate content, and ignore if the content is same. Just to confirm, is my assumption wrong?

Reply:
> I am a bit confused why we are doing a duplicate content check here. Is the main purpose to avoid updating the distributor if the contents are same?

The only purpose of doing the duplicate content check it avoid parsing the cert using a call to `tls.LoadX509KeyPair` and updating the `Distributor`. The `Distributor` in Go provides a `fetch` style API. So, even if we update it with the same key material, nothing will break. So, this does not affect correctness and is not intended to be a performance optimization. If you feel strongly against storing the file contents and doing this check, I can get rid of it. Let me know.

> I am not sure if this performance check is worth to do, because before, we will just do one read from disk for each files, and now we will need to do twice.

Why do you say we are reading the file twice? Is it because the `tls.LoadX509KeyPair` will read it again? Would it make sense to use [tls.X509KeyPair](https://golang.org/pkg/crypto/tls/#X509KeyPair) in that case, which directly accepts the PEM contents.

> Also, I was under the impression that the distributor will check the duplicate content, and ignore if the content is same. Just to confirm, is my assumption wrong?

No, the `Distributor` does not do this.

> It might be better if we can check the cert we got from tls.LoadX509KeyPair directly. Is it possible?

The `tls.Certificate` type does not define an `Equal` method. The `x509.Certificate` does, but there is an issue that the `Equal` method does not check for `nil` and will panic if one of the arguments is `nil`. This has been fixed in `go1.14`, but we do support older Go versions. Even if we only support `go1.14`, the documentation for [tls.LoadX509KeyPair](https://golang.org/pkg/crypto/tls/#LoadX509KeyPair) says that the `Leaf` field will be left `nil`. So, I'm not sure if its going to worth the effort.

Reply:
Switched it to `tls.X509KeyPair` to save one file read.